### ISSUE TITLE
zpool destroy fails on zpool with suspended IO

### ISSUE THREAD

**Issue by kiranfractal** [openzfs/zfs#2878]
Hi,

I am trying disk replacement from a single disk zpool and it results to suspended IO and does not allow me to destroy the zpool.
## Steps to reproduce:
1. create a zpool with single disk (zpool create zp1 /dev/sda)
2. remove the disk 
3. zpool status shows disk unavailable
4. insert a new disk
5. zpool replace zp1 /dev/sda /dev/sdb
6. it says "cannot replace /dev/sda with /dev/sdb : pool I/O is currently suspended"

The above steps holds good for zpool destroy as well.

Now the zpool is unusable and only option for me to destroy zpool is to reboot the system and destroy the old pool and create a new pool.

Is there any other option can I try so that I can replace the disk without rebooting the system ?

Thanks,
Kiran.

**Comment by behlendorf** [comment]
I've merged 87a63dd to master which ensures a reasonable error is generated when attempting to destroy or export a suspended pool.  This is far preferable to a hang.  However, I'd still like to leave this issue open so we can explore what is reasonable behavior for a force option.

87a63dd Prevent "zpool destroy|export" when suspended

**Comment by darkpixel** [comment]
We are trying to use ZFS for the exact same scenario.  Our clients have a pool of spinning rust they use for primary data storage.  We zfs send that data to two places--an off-site backup, and a local single external USB drive.

Basically they want *fast* (local) recovery options in case somehow the primary data storage dies from the local USB drive.  And they want an off-site backup in case the office burns down.

We have external USB drives (Seagate) drives at 24 locations and it seems like one dies every month.  When they die, we ship a new drive out, get it plugged in, and then have to reboot the box because we can't destroy 'backup-pool' because IO is suspended.  Once the box has been rebooted 'backup-pool' no longer shows up, then we can create a new 'backup-pool' from the newly installed drive.

I agree with @kiranfractal that there should be a '-F' option (in addition to '-f') that would force removal of the pool regardless of IO being suspended.

**Comment by kernelOfTruth** [comment]
referencing

https://www.illumos.org/issues/4128 disks in zpools never go away when pulled
https://github.com/illumos/illumos-gate/commit/39cddb1

**Comment by ashjas** [comment]
This same issue i am facing in such a scenario:
1) sudo zpool create mypool -m ~/ZFS  mirror /dev/sdc1 /dev/sdc2
I am using mirror on a single disk... i know this is inefficient.. but im testing...

2) sudo zfs mount / umount + sudo zpool import/export works as expected.

3) reboot

4) put in a flash drive, so that my zpool device become as : /dev/sdd1 & /dev/sdd2.  // sdc1 is now for the inserted flashdrive.

5) neither sudo zfs mount / umount  nor  sudo zpool import/export works as expected.


----------> output of commands issued:
```
sudo zpool status -v
  pool: mypool
 state: ONLINE
  scan: scrub repaired 0 in 0h0m with 0 errors on Sun Jan 31 19:12:37 2016
config:

	NAME                                                 STATE     READ WRITE CKSUM
	mypool                                               ONLINE       0     0     0
	  mirror-0                                           ONLINE       0     0     0
	    ata-WDC_WD1600BEVS-00UST0_WD-WXEY07V22050-part1  ONLINE       0     0     0
	    ata-WDC_WD1600BEVS-00UST0_WD-WXEY07V22050-part2  ONLINE       0     0     0

errors: No known data errors
```

```
sudo zpool import
no pools available to import

```
```
sudo zpool import mypool
cannot import 'mypool': a pool with that name already exists
use the form 'zpool import <pool | id> <newpool>' to give it a new name
```


```
sudo zpool export mypool
umount: /home/ashish/ZFS: not mounted
cannot unmount '/home/ashish/ZFS': umount failed
```

```
sudo zfs mount mypool
cannot mount 'mypool': filesystem already mounted
```

```
sudo zfs get all mypool | grep mount
mypool  mounted               yes                    -
mypool  mountpoint            /home/ashish/ZFS       local
mypool  canmount              on                     default
```

Surprisingly, issuing a scrub command:
`sudo zpool scrub mypool`

correctly tries to scan the Usb HDD, as i can see its LED blinking.

Also, if in this booted session of ubunt, if i fix the path back to /dev/sdc1 /dev/sdc2 for the zpools, by removing the flash drive , and reattaching the zpool drive, the above does not change.. the issue still remains. This only gets fixed if i reboot the system.

[dmesg.log](https://gist.github.com/ashjas/d5ec0f278966f6ef1ae1)

**Comment by ilovezfs** [comment]
@ashjas I'm not seeing anything wrong or unexpected there, and certainly nothing having to do with device names. The only thing that looks like a problem is
```
cannot unmount '/home/ashish/ZFS': umount failed
```
which typically indicates that that particular dataset is busy. Often that can be resolved with -f or closing whatever files are opened. lsof is your friend.

**Comment by ashjas** [comment]
@ilovezfs
For the first scenario where the device name get changed...
5) neither sudo zfs mount / umount nor sudo zpool import/export works as expected.

This is not an issue? I'm unable to mount the volume.
If I'm giving wrong commands ,let me know. But how this isn't an issue is out of my head.

Secondly -f doesn't make any difference at all!

**Comment by darkpixel** [comment]
@ashjas: I don't create pools like this: ```sudo zpool create mypool -m ~/ZFS mirror /dev/sdc1 /dev/sdc2```

I don't know about how ZFS handles that, but /dev/sda, /dev/sdb, etc... can change when devices are changed.

I usually create the pools by using devices in /dev/disk/by-id/*

Those ID numbers are not supposed to change, even if devices are switched around.

**Comment by prushkaru** [comment]
I come here becouse of my case with ZFS: this is realy strange to use one disk per ZFS pool =) Yes. But i did it with USB storage with hope to make IO faster (1Tb of data). So suddenly after all i have to give away this disk -- power off storage and hard unlink it outta system. So what i got:
cannot open 'snaproot': pool I/O is currently suspended
=)
And no more way to destroy it... This is the case.
Yes, i will reboot the system now and look what happen then =) to be continued...

**Comment by bjquinn** [comment]
Yeah, I run into this all the time.  Gotten to the point that this one issue is the cause of 99% of all reboots for my Linux servers.

**Comment by prushkaru** [comment]
I am sorry becouse of my lost in 18 Feb 2016 =))) Yes. All was well done afte reboot. =) Excuse me. Thanks to all. Best wishes

**Comment by GregorKopka** [comment]
> Let me ask an opened ended question. For a pool which is suspended what would you expect the
> following behavior to be for the following commands. Where -f mean force and -F means hardforce.

> 1.  zpool export|destroy < pool >

Application: 
Normally working pool

Function:
Work as-is
Always fail with a sane error message (for whatever reason) instead of getting stuck

> 2.  zpool export|destroy -f < pool >

Application:
Problems with upper layers (locks from applications etc)

Function:
Execute 1, in case of error invalidate all open file handles to pool datasets and terminate all in-flight I/O on them with I/O errors, detach them from the datasets, sync, then re-execute 1 to now being able to unmount the datasets. The now stale file handles need to be held by zfs until closed so the file handle interface to the upper layer dosn't get into undefined state.
Always fail with a sane error message (for whatever reason) instead of getting stuck

> 3.  zpool export|destroy -F < pool >

Application:
Problems with lower layer (disk failures//hang/missing, suspended pool)

Function:
Execute 2, in case of error invalidate and terminate all outstanding I/O to all pool vdevs, drop all buffers of pool (dirty or not), invalidate and free all data structures associated with the pool (except any dangling stale file handles from 2 which need to be kept till the applications close them), in case of destroy an suspended pool: skip the destroy part (since pool members can't be written to)
Always return succesfully and complete, never get stuck.

**Comment by bjquinn** [comment]
@GregorKopka I think that behavior sounds pretty reasonable.  I can see the use case for both -f and -F.  My particular use case lately has been your description of the behavior of -F.

**Comment by GregorKopka** [comment]
Having throught about it a bit longer, 2) should also hard terminate any I/O originating from zfs (command) operations on the pool (snapshot, list, send/recv, destroy, ...) since the pool is to be forcible exported/destroyed.

Brian asked open ended, so my answer is in general and not only for suspended pools - which would always need -F in my scheme since they can't be written to, so 1 will always fail and through that also 2 since it relies on 1 to work on the second run after (cleaning up any blocks from the upper layers).

**Comment by GregorKopka** [comment]
@behlendorf Please reconsider stuck zpool/zfs command / inability to remove a pool from the system *not* being a bug.

This scenario isn't that uncommon with USB-disk based pools used for backups, 'reboot to fix it' should keep to be a speciality of windows...

**Comment by bjquinn** [comment]
Agreed.  USB pools for backups are not reliably usable given this issue.

**Comment by behlendorf** [comment]
@GregorKopka I agree.  There's lots of room for improvement when someone has the time to focus on this.  I've just removed the Bug tag from all issues because it wasn't actually helpful.

**Comment by mailinglists35** [comment]
@behlendorf is there any hope for someone to implement this anytime soon?

**Comment by behlendorf** [comment]
No developers I know of are currently working on this.

**Comment by lnxbil** [comment]
I just ran into the same problem after accidentially removing the only disk the pool consists of via a `echo 1 > /sys/block/sdr/delete` call and I was (after a lot of different tries) able to reimport the pool after reading this thread:

```
$ zpool status -v externes-backup
  pool: externes-backup
 state: ONLINE
status: One or more devices are faulted in response to IO failures.
action: Make sure the affected devices are connected, then run 'zpool clear'.
   see: http://zfsonlinux.org/msg/ZFS-8000-HC
  scan: scrub repaired 0B in 29h48m with 0 errors on Thu Aug  4 20:51:21 2016
config:

        NAME                     STATE     READ WRITE CKSUM
        externes-backup          ONLINE       1     9     0
          externes_backup_crypt  ONLINE       1     0     0

errors: List of errors unavailable: pool I/O is currently suspended
```

Just for others that might run into the same problem:
* I have/had a luks-crypted device called `externes_backup_crypt`
* Closing luks device (to be able to have the same name again, otherwise it yields `no such device in pool`)

        $ cryptsetup luksClose externes_backup_crypt

* Rescan SCSI-Bus

        $ for bus in /sys/class/scsi_host/host*; do echo "- - -" > $bus/scan; done

* Decrypt device

        $ cryptsetup luksOpen /dev/sds externes_backup_crypt

* put disk online

        $ zpool online externes-backup  /dev/mapper/externes_backup_crypt
    cannot online /dev/mapper/externes_backup_crypt: pool I/O is currently suspended

* not, you have to clear the pool and it'll continue to work again:

        $ zpool clear externes-backup

* Afterwards, the pool is imported and automatically scrubbed:

```
$ root@backup ~ > zpool status -v externes-backup
  pool: externes-backup
 state: ONLINE
status: One or more devices has experienced an error resulting in data
        corruption.  Applications may be affected.
action: Restore the file in question if possible.  Otherwise restore the
        entire pool from backup.
   see: http://zfsonlinux.org/msg/ZFS-8000-8A
  scan: scrub in progress since Wed Feb 14 10:32:32 2018
        4,87G scanned out of 6,14T at 12,1M/s, 147h47m to go
        0B repaired, 0,08% done
config:

        NAME                     STATE     READ WRITE CKSUM
        externes-backup          ONLINE       0     0     0
          externes_backup_crypt  ONLINE       0     0     0

errors: Permanent errors have been detected in the following files:

        <metadata>:<0x0>
        <metadata>:<0x1>
        <metadata>:<0x18675>
        <metadata>:<0x3dca1>
```

Now, I'm at least able to export the pool correctly.

**Comment by felisucoibi** [comment]
@lnxbil thanks, at least we have something,i have the sameproblem, using LUKS+ZFS and when something goes wrong i have to restart the server, the problem comes when i'm not at home and i cant reboot.

**Comment by kpande** [comment]
@dweeezil can you give some insight here into the issue with suspended single disk pools? I recall you were working on the ability to forcibly unload a pool from memory.

**Comment by kpande** [comment]
#5242 is the main point of communication for this issue.

**Comment by GregorKopka** [comment]
@behlendorf this issue was closed without a _who_ and _when_ -  is that supposed to happen?

There (AFAIK) still isn't a solution to this (only workarounds for a subset of the problem), so the defect remains. Also quite a lot of information here that would certainly be interesting to someone who decides to work on this at some point in the future. 

Please repoen.

**Comment by behlendorf** [comment]
It's not clear to me exactly how this was closed.  I don't have any objection to reopening it since this is still and issue.

**Comment by devZer0** [comment]
this is still an issue in 2020

i had some flaky disks in a server and wanted to remove those, i'm somewhat sure i did zpool export or destroy before removing those 3 disks, but i must have done wrong, at least the pool was online and mounted while the disks being ripped off (didn't double check before) - now the pool it's in suspended state and apparently , it's not possible to destroy/remove from the system when in this state.

as there are VMs online on this system, i'm rather curious what to do without shutting them all down and reboot....

i'm getting this in dmesg after another "zpool destroy..." hung for a while....

[1510610.643638] INFO: task txg_sync:1714 blocked for more than 120 seconds.
[1510610.643684]       Tainted: P          IOE     5.4.34-1-pve #1
[1510610.643711] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[1510610.643741] txg_sync        D    0  1714      2 0x80004000
[1510610.643744] Call Trace:
[1510610.643752]  __schedule+0x2e6/0x700
[1510610.643754]  schedule+0x33/0xa0
[1510610.643757]  schedule_timeout+0x152/0x300
[1510610.643761]  ? __next_timer_interrupt+0xd0/0xd0
[1510610.643763]  io_schedule_timeout+0x1e/0x50
[1510610.643772]  __cv_timedwait_common+0x12f/0x170 [spl]
[1510610.643774]  ? wait_woken+0x80/0x80
[1510610.643779]  __cv_timedwait_io+0x19/0x20 [spl]
[1510610.643851]  zio_wait+0x13a/0x280 [zfs]
[1510610.643902]  dsl_pool_sync+0x46e/0x500 [zfs]
[1510610.643958]  spa_sync+0x5b2/0xfc0 [zfs]
[1510610.644014]  ? spa_txg_history_init_io+0x106/0x110 [zfs]
[1510610.644077]  txg_sync_thread+0x2d9/0x4c0 [zfs]
[1510610.644156]  ? txg_thread_exit.isra.12+0x60/0x60 [zfs]
[1510610.644163]  thread_generic_wrapper+0x74/0x90 [spl]
[1510610.644167]  kthread+0x120/0x140
[1510610.644173]  ? __thread_exit+0x20/0x20 [spl]
[1510610.644175]  ? kthread_park+0x90/0x90
[1510610.644178]  ret_from_fork+0x35/0x40
[1510610.644182] INFO: task zed:2736 blocked for more than 120 seconds.
[1510610.644208]       Tainted: P          IOE     5.4.34-1-pve #1
[1510610.644231] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[1510610.644260] zed             D    0  2736      1 0x00004000
[1510610.644262] Call Trace:
[1510610.644264]  __schedule+0x2e6/0x700
[1510610.644266]  schedule+0x33/0xa0
[1510610.644267]  io_schedule+0x16/0x40
[1510610.644272]  cv_wait_common+0xb5/0x130 [spl]
[1510610.644274]  ? wait_woken+0x80/0x80
[1510610.644279]  __cv_wait_io+0x18/0x20 [spl]
[1510610.644336]  txg_wait_synced_impl+0xc9/0x110 [zfs]
[1510610.644393]  txg_wait_synced+0x10/0x40 [zfs]
[1510610.644449]  spa_vdev_state_exit+0x8a/0x160 [zfs]
[1510610.644505]  vdev_online+0x2bb/0x3e0 [zfs]
[1510610.644564]  zfs_ioc_vdev_set_state+0x86/0x190 [zfs]
[1510610.644622]  zfsdev_ioctl+0x6db/0x8f0 [zfs]
[1510610.644626]  ? lru_cache_add_active_or_unevictable+0x39/0xb0
[1510610.644629]  do_vfs_ioctl+0xa9/0x640
[1510610.644632]  ? handle_mm_fault+0xc9/0x1f0
[1510610.644634]  ksys_ioctl+0x67/0x90
[1510610.644636]  __x64_sys_ioctl+0x1a/0x20
[1510610.644639]  do_syscall_64+0x57/0x190
[1510610.644642]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[1510610.644644] RIP: 0033:0x7f9c54306427
[1510610.644649] Code: Bad RIP value.
[1510610.644650] RSP: 002b:00007f9c534a1078 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[1510610.644652] RAX: ffffffffffffffda RBX: 00007f9c534a10c0 RCX: 00007f9c54306427
[1510610.644653] RDX: 00007f9c534a10c0 RSI: 0000000000005a0d RDI: 0000000000000009
[1510610.644654] RBP: 00007f9c534a5ab0 R08: 00000000b5c9ecf4 R09: 00007f9c546bb8e6
[1510610.644655] R10: 00007f9c44041450 R11: 0000000000000246 R12: 00007f9c44015740
[1510610.644656] R13: 00007f9c4401e980 R14: 00007f9c534a6b40 R15: 00007f9c534a4670
[1510731.477176] INFO: task zed:2736 blocked for more than 241 seconds.
[1510731.477210]       Tainted: P          IOE     5.4.34-1-pve #1
[1510731.477232] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
